### PR TITLE
Reimplement PO parsing to handle msgstrs with spaces

### PR DESCRIPTION
**Pull request by kbairak**
Problem and/or solution
-----------------------
Handle '\n' as wildcard (empty/non empty) depending on the whole file context reduce errors on file handling. 

```
1st msgstr  | nth msgstr  | result  | notes
------------| ----------- | ------- | --
''          | ''          | success |  
''          | '\n'        | success | '\n' acts as empty
''          | 'non-empty' | error   |  
'\n'        | ''          | success | '\n' acts as empty
'\n'        | '\n'        | success | ‘\n’ acts as empty
'\n'        | 'non-empty' | success | '\n' acts as not empty
'non-empty' | ''          | error   |  
'non-empty' | '\n'        | success | '\n' acts as non empty (this is the one that changed)
'non-empty' | 'non-empty' | success |  
```

How to test
-----------
1. Open testbed
2. Test above combinations by using the template
```
msgctxt "Number of days"
msgid "%s day"
msgid_plural "%s days"
msgstr[0] "\n"
msgstr[1] "\n"

#, python-format-2
msgctxt "Number of numbers"
msgid "%s number"
msgid_plural "%s numbers"
msgstr[0] "\n"
msgstr[1] "\n"
```

Reviewer checklist
------------------

Code:
* [ ] Change is covered by unit-tests
* [ ] Code is well documented, well styled and is following [best practices](https://tem.transifex.com)
* [ ] Performance issues have been taken under consideration
* [ ] Errors and other edge-cases are handled properly

PR:
* [ ] Problem and/or solution are well-explained
* [ ] Commits have been squashed so that each one has a clear purpose
* [ ] Commits have a proper commit message [according to TEM](https://tem.transifex.com/github-guide.html#working-on-a-feature)

### CODE REVIEW

File: openformats/formats/po.py [hunk 1]
```diff
@@ -21,283 +20,237 @@ class PoHandler(Handler):
...
-            else:
-                raise ParseError(
-                    u"The entry with msgid `{}` includes an empty msgstr. "
-                    u"Provide a value and try again.".format(entry.msgid)
+    def _get_metadata(self, entry):
+        return {
+            "occurrences": ", ".join(
+                (
+                    ":".join((item for item in occurence))
+                    for occurence in entry.occurrences
```
Comment (fathineos) [inline openformats/formats/po.py:221]: super minor: typo double `rr` needed

File: openformats/formats/po.py [hunk 2]
```diff
@@ -361,18 +314,18 @@ def _format_occurrences(occurrences):
         The occurrences is a list of 2-tuples that contain the file and the
         line number.
         """
-        return ', '.join([
-            ':'.join([
-                item for item in occurence
-            ]) for occurence in occurrences
-        ])
+        return ", ".join(
+            [":".join([item for item in occurence]) for occurence in occurrences]
```
Comment (fathineos) [inline openformats/formats/po.py:318]: super minor: typo double `rr` needed

File: openformats/formats/po.py [hunk 3]
```diff
@@ -21,283 +20,237 @@ class PoHandler(Handler):
...
+                            "'{}' and msgid_plural '{}'"
+                        ).format(entry.msgid, entry.msgid_plural)
+                    )
+            elif all((not string for string in msgstrs.values())):
+                string_type = "EMPTY"
+            else:
+                string_type = "SPACES"
+
+            if is_source:
+                # - If all entries are empty, then this is a POT file and we shouldn't
```
Comment (fathineos) [inline openformats/formats/po.py:103]: useful comment

Comment (fathineos) [reply openformats/formats/po.py:103]: I think that it would be helpful to add a var declaring if the file is considered to be `po` or `pot`.

File: openformats/formats/po.py [hunk 4]
```diff
@@ -21,283 +20,237 @@ class PoHandler(Handler):
...
+                    # Translation files are all assumed to be PO files. Empty entries
+                    # should be removed from the stringset
+                    new_po.remove(entry)
+                    continue
+                string_values = string_data["msgstrs"]
+            openstring = OpenString(
+                string_data["key"],
+                string_values,
+                fuzzy=string_data["fuzzy"],
+                order=next(order),
```
Comment (fathineos) [inline openformats/formats/po.py:163]: I think that the order will defer from prior implementation. The `fuzzy` entries were skipped.

**Review by fathineos — CHANGES_REQUESTED** [review]

File: openformats/formats/po.py [hunk 5]
```diff
@@ -21,283 +20,237 @@ class PoHandler(Handler):
...
+                            "with msgid '{}' and msgid_plural '{}'"
+                        ).format(entry.msgid, entry.msgid_plural)
+                    )
+                key = ":".join((msgid, msgid_plural))
             else:
-                result.append('# %s' % header)
-        result.append(six.text_type(po_file.metadata_as_entry()))
-        for entry in po_file:
-            result.append(six.text_type(entry))
+                if any((value.strip() for value in entry.msgstr_plural.values())):
```
Comment (fathineos) [inline openformats/formats/po.py:57]: This was `entry.msgstr_plural` on the previous implementation. I think that it was correct, should we allow spaces if provided in a non pluralized string?

File: openformats/formats/po.py [hunk 6]
```diff
@@ -21,283 +20,237 @@ class PoHandler(Handler):
...
+        order = itertools.count()
+
+        for string_data, entry in zip(strings, list(new_po)):
+            if is_source:
+                if "EMPTY" in string_types:
+                    # This is a POT file, we must consider the msgids as strings
+                    if string_data["pluralized"]:
+                        string_values = {
+                            0: string_data["msgid"],
+                            1: string_data["msgid_plural"],
```
Comment (fathineos) [inline openformats/formats/po.py:142]: not relevant to this pull request: Why we explicitely set the `0` and `1` for this pluralized string?

Comment (kbairak) [reply openformats/formats/po.py:142]: Because PO files don't follow the indexes we use. When an entry in a PO file has msgstr_plural, the indexes are 0 and 1, not 1 and 5. This means that if we want to use the correct indexes in openformats, we would have to provide the language's plural rules as an argument to `parse` and `compile`, which we decided not to do. So, we pretend that the rules are 0 and 1 (0, 1 and 2 if there are 3 plurals etc) and fix them during postprocessing.

File: openformats/formats/po.py (hunk 4 above)
Comment (kbairak) [reply openformats/formats/po.py:163]: Will look into it.

Comment (kbairak) [reply openformats/formats/po.py:163]: Good catch, it would have slipped my radar. Thanks!

**Comment by fathineos** [comment]
Please add some pull request description & rebase/squash and we are good to go!

**Review by fathineos — APPROVED** [review]
LGTM
